### DISCUSSION TITLE
Versioning with a History Table in 2.0 and __table__ attribute

### DISCUSSION THREAD

**Discussion by PHvL** [sqlalchemy/sqlalchemy#9546] [Usage Questions] [answered]
In 1.4 I'm using the `Versioned` class from history_meta, and I'm trying to migrate these models to 2.0.

The models defined with the [2.0 version of history_meta](https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.versioned_history) seem to work fine on their own, but I get an exception, if I try to access the `__table__` (e.g. for attaching events), try to use `sqlalchemy.inspect`, or try to use the `Versioned` object in a join. In 1.4 the `Versioned`-classes behave just like any other `Base`-class.

```python
from sqlalchemy import join, ForeignKey, Column, Integer, String
from sqlalchemy.orm import DeclarativeBase, column_property
from history_meta import Versioned
class Base(DeclarativeBase):
        pass

class ThingExternal(Versioned, Base):
        __tablename__ = 'things_external'
        id = Column(Integer, primary_key=True, autoincrement=False)
        external_attribute = Column(String)

class ThingLocal(Versioned, Base):
        __tablename__ = 'things_local'
        id = Column(Integer, ForeignKey(ThingExternal.id), primary_key=True, autoincrement=False)
        internal_attribute = Column(String)
```

Then `ThingExternal.__table__` fails with 
> `AttributeError: type object 'ThingExternal' has no attribute '__table__'`

And
```python
class Thing(Base):
        __table__ = join(ThingExternal, ThingLocal)
        id = column_property(ThingExternal.id, ThingLocal.id)
```
fails with
> `sqlalchemy.exc.ArgumentError: FROM expression, such as a Table or alias() object expected, got <class '__main__.ThingExternal'>.`

For context: we sync information of things from an external source into things_external, add information controlled by us in things_local and access the combined information in one joined object Thing.

I would be grateful for any hints on what I did wrong, migrating our current code to SQLAlchemy 2.0.

**Comment by zzzeek** [comment]
I had to port Versioned to 2.0 after the fact, and it was not that easy to do for whatever reason, so these are likely more regressions in the example.  I will have to fix them individually.

**Reply by zzzeek** [reply, marked as answer]
it looks like DeclarativeBase superclass is the issue.    if you use declarative_base() for now everything seems to be the same.

**Reply by PHvL** [reply]
> it looks like DeclarativeBase superclass is the issue. if you use declarative_base() for now everything seems to be the same.

Thanks! Yes that works.

Thank you for your super-fast reply and your work!
